app/controllers/admin_controller.rb

```diff
@@ -1,8 +1,12 @@
 # frozen_string_literal: true
 
+#  ==== Admin utilities
+#  test_flash_redirection::      <tt>(R . .)</tt>
+
 class AdminController < ApplicationController
   before_action :login_required
```

Comment:
Why not just in-line the RestrictAccesstoAdminMode module here so all admin (sub)controllers inherit it by default?

Reply:
That was my original idea. Tried it - I think something in the `DonationsController` does not work with `RestrictAccesstoAdminMode`. I tried it in each controller independently.

Reply:
Oh- I remember. `DonationsController` has different redirects, if you're not an admin, than `RestrictAccesstoAdminMode`.

Reply:
Looks like /admin/donations has the same redirect problem, and again, it would be totally fine to have it redirect to / instead of /support/donate.

Reply:
Ohhh, yeah. The blocker for having `RestrictAccesstoAdminMode` everywhere in Admin is... `TurnAdminOn`. You can't restrict access to that! 

That's why I ended up leaving `Admin::UsersController` and `Admin::DonationsController` with those polite custom tailored redirects. 

One solution. I could do a separate top level controller that doesn't have restricted access, called `AdminAccessController`. Then you'd hit the show action of that controller at `/admin-access/` with your `turn_on`, `turn_off`, or `drop_out` param. Sound good?

Let me know and I will reconfigure this. I like the potential simplification of the `AdminController`.

Reply:
Am i correct in thinking that the only "turn_admin_on" button is in the top bar [A] button ?

And that the sidebar only ever shows a "turn_admin_off" link (to admins)?

Reply:
Oh hey, i've got a great name for this controller. `AdminModeController`

Reply:
Okay, one final item. The `AdminController` method `test_flash_redirection` cannot require admin access. Where should I put that...?

Reply:
re: turn_admin_on require nonadmin access -- of course!  thanks.

re: admin_mode_controller -- excellent!  Good job sticking with it until the solution appeared!

re: test_flash_redirection -- test controller??  Oooh!  There would be a test_controller_test.rb :)

Reply:
Hahaha `test_controller_test.rb` ok

One more action requires special handling... `switch_users`. 
It's got an extra OR condition governing access to the form, `session[:real_user_id].blank?`

Idea: I can put the `switch_user` form in the `AdminModeController` (as `edit`?) to keep the conditions of `AdminController` the same.

Reply:
If that works, sounds good to me.  creative.

Reply:
It turns out that controllers namespaced under Admin like `Admin::DonationsController` do not automatically inherit anything from `AdminController` unless they explicitly inherit from that controller class, instead of `ApplicationController`. (of course!) 

So I can move the admin mode controller back under `Admin::ModeController` and just have it inherit from `ApplicationController` instead of `AdminController`. 'doh

Reply:
Haha!  I didn't think of that either!

Reply:
...But i'm not going to move it. It's fine as `AdminModeController`

Reply:
That's great.  Thanks!